src/counter.cc

```diff
@@ -0,0 +1,327 @@
+// Copyright 2015 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "benchmark/benchmark.h"
+#include "check.h"
+#include "internal_macros.h"
+
+#include <cstring>
+
+namespace benchmark {
+
+Counter::Counter()
+  : name_buf_(), name_mem_(nullptr), mem_size_(0), name_(nullptr), value_(0.), flags_(kDefaults) {
+  name_buf_[0] = '\0';
+}
+
+Counter::Counter(const char* name)
+  : name_buf_(), name_mem_(nullptr), mem_size_(0), name_(nullptr), value_(0.), flags_(kDefaults) {
+  _SetName(name);
+}
+
+Counter::Counter(const char* name, double v)
+  : name_buf_(), name_mem_(nullptr), mem_size_(0), name_(nullptr), value_(v), flags_(kDefaults) {
+  _SetName(name);
+}
+
+Counter::Counter(const char* name, double v, uint32_t f)
+  : name_buf_(), name_mem_(nullptr), mem_size_(0), name_(nullptr), value_(v), flags_(f) {
+  _SetName(name);
+}
+
+Counter::Counter(Counter const& that)
+  : name_buf_(), name_mem_(nullptr), mem_size_(0), name_(nullptr), value_(that.value_), flags_(that.flags_) {
+  _SetName(that.name_);
+}
+Counter& Counter::operator= (Counter const& that) {
+  if(&that == this) return *this;
+  _SetName(that.name_);
+  value_ = that.value_;
+  flags_ = that.flags_;
+  return *this;
+}
+
+#ifdef BENCHMARK_HAS_CXX11
+Counter::Counter(Counter && that)
+  : name_buf_{0}, name_mem_(that.name_mem_), mem_size_(that.mem_size_), name_(that.name_), value_(that.value_), flags_(that.flags_) {
+  memcpy(name_buf_, that.name_buf_, sizeof(name_buf_));
+  that.name_mem_ = nullptr;
+  that.mem_size_ = 0;
+  that.name_ = nullptr;
+}
+Counter& Counter::operator= (Counter && that) {
```

Comment:
the default operator isn't good enough? if you switch to string it should be, right?